### PR TITLE
CI: add a script for dynamically computing CI job matrix

### PR DESCRIPTION
**Pull request by Kobzol**
It would be great if was easier to run specific CI workflows locally, and also to allow us to spawn a specific CI workflow by bors, to enable running arbitrary try builds. See discussion [here](https://rust-lang.zulipchat.com/#narrow/stream/242791-t-infra/topic/CI.20workflows.20refactoring).

This PR is a first step in that direction.
- Moves the definition of CI runners and (for now) PR jobs into a separate `jobs.yml` file.
- Adds a simple Python script that reads the file, decides which jobs should be active for the current CI workflow, and prints them as JSON to their output.
- The PR job then reads this output and generates its job matrix based on it.

By moving the job definitions from `ci.yml` into a separate file, we can handle it programmatically, which should make it easier to both do local execution of CI jobs and also to do arbitrary try builds.

### CODE REVIEW

File: src/ci/github-actions/ci.yml [hunk 1]
```diff
@@ -340,30 +340,29 @@ concurrency:
...
-            <<: *job-linux-4c
-
-          - name: x86_64-gnu-llvm-17
-            env:
-              ENABLE_GCC_CODEGEN: "1"
-            <<: *job-linux-16c
-
-          - name: x86_64-gnu-tools
-            <<: *job-linux-16c
+        include: ${{ fromJSON(needs.calculate_matrix.outputs.jobs) }}
```
Comment (pietroalbini) [inline src/ci/github-actions/ci.yml:365]: We should add a comment here explained where those jobs are defined now (the `jobs.yml` file), to make sure people wanting to add new jobs to the PR builder know where the configuration has been moved.

File: src/ci/github-actions/jobs.yml [hunk 2]
```diff
@@ -0,0 +1,46 @@
+x--expand-yaml-anchors--remove:
```
Comment (pietroalbini) [inline src/ci/github-actions/jobs.yml:1]: We should add a comment here saying this file does **not** need to anchors to be expanded when you change it. This way, people used to run the command every time they changed the CI config don't run it needlessly.

**Review by emilyalbini — COMMENTED** [review]
r=me after applying the suggestions

**Comment by Kobzol** [comment]
Added comments.

**Comment by lqd** [comment]
BTW is the procedure described in https://rustc-dev-guide.rust-lang.org/tests/ci.html#using-ci-to-test still recommended? Or should we rather change `jobs.toml` to temporarily enable PR builders?

**Comment by Kobzol** [comment]
Currently `jobs.yml` needs to be modified to change what is executed on PRs, yeah. I plan to move the rest of the jobs to `jobs.yml` soon, then this section should be rewritten.

**Comment by RalfJung** [comment]
Wow, I had no idea this is possible. Does that mean we can entirely get rid of the `./x.py run` step to generate the yml files, and just compute them on-the-fly instead?

Also, this did had some probably unintended side-effect -- the actually relevant jobs for a PR no longer show up at the top of the list, half of them are hidden behind the scrollbar:

image

If there's a way to order these `CI / {auto,try,master}` down again, that'd be nice. :D

**Pull request merged**